File: src/parser.rs
use crate::ast::*;
use crate::error::CompileError;
use crate::lexer::Token;

pub struct Parser {
    tokens: Vec<Token>,
    pos: usize,
}

impl Parser {
    pub fn new(tokens: Vec<Token>) -> Self {
        Parser { tokens, pos: 0 }
    }

    fn peek(&self) -> &Token {
        &self.tokens[self.pos]
    }
    fn eat(&mut self) {
        if self.pos < self.tokens.len() {
            self.pos += 1
        }
    }
    fn expect(&mut self, expected: Token) -> Result<(), CompileError> {
        if *self.peek() == expected {
            self.eat();
            Ok(())
        } else {
            Err(CompileError::Parse(format!(
                "Expected {:?}, found {:?}",
                expected,
                self.peek()
            )))
        }
    }

    /// Top‐level entry: parse a whole program.
    pub fn parse_program(&mut self) -> Result<Program, CompileError> {
        let mut funcs = Vec::new();
        let mut stmts = Vec::new();

        while *self.peek() != Token::EOF {
            if *self.peek() == Token::Fn {
                funcs.push(self.parse_function()?);
            } else {
                let stmt = self.parse_statement()?;
                stmts.push(stmt);
                if *self.peek() == Token::Semicolon {
                    self.eat();
                }
            }
        }

        Ok(Program {
            functions: funcs,
            statements: stmts,
        })
    }

    /// Parse `fn name(arg1, arg2, …) { … }`
    fn parse_function(&mut self) -> Result<Function, CompileError> {
        self.expect(Token::Fn)?;
        let name = match self.peek() {
            Token::Ident(n) => n.clone(),
            _ => return Err(CompileError::Parse("Expected function name".into())),
        };
        self.eat();
        self.expect(Token::LParen)?;
        let mut params = Vec::new();
        if *self.peek() != Token::RParen {
            loop {
                if let Token::Ident(n) = self.peek() {
                    params.push(n.clone());
                    self.eat();
                } else {
                    return Err(CompileError::Parse("Expected parameter name".into()));
                }
                if *self.peek() == Token::Comma {
                    self.eat();
                    continue;
                }
                break;
            }
        }
        self.expect(Token::RParen)?;
        let body = self.parse_block()?;
        Ok(Function { name, params, body })
    }

    /// Parse a `{ stmt; stmt; … }` block
    fn parse_block(&mut self) -> Result<Vec<Statement>, CompileError> {
        self.expect(Token::LBrace)?;
        let mut v = Vec::new();
        while *self.peek() != Token::RBrace {
            let stmt = self.parse_statement()?;
            self.expect(Token::Semicolon)?;
            v.push(stmt);
        }
        self.expect(Token::RBrace)?;
        Ok(v)
    }

    /// Parse any single statement (var, assign, if, while, return, print, expr‐stmt).
    fn parse_statement(&mut self) -> Result<Statement, CompileError> {
        match self.peek() {
            Token::Var => {
                self.eat();
                let name = if let Token::Ident(n) = self.peek() {
                    n.clone()
                } else {
                    return Err(CompileError::Parse("Expected var name".into()));
                };
                self.eat();
                self.expect(Token::Eq)?;
                let expr = self.parse_expr()?;
                Ok(Statement::VarDecl { name, expr })
            }
            Token::Let => {
                self.eat();
                let name = if let Token::Ident(n) = self.peek() {
                    n.clone()
                } else {
                    return Err(CompileError::Parse("Expected let name".into()));
                };
                self.eat();
                self.expect(Token::Eq)?;
                let expr = self.parse_expr()?;
                Ok(Statement::LetDecl { name, expr })
            }
            Token::If => {
                self.eat();
                self.expect(Token::LParen)?;
                let cond = self.parse_expr()?;
                self.expect(Token::RParen)?;
                let then_branch = self.parse_block()?;
                let else_branch = if *self.peek() == Token::Else {
                    self.eat();
                    Some(self.parse_block()?)
                } else {
                    None
                };
                Ok(Statement::If {
                    cond,
                    then_branch,
                    else_branch,
                })
            }
            Token::While => {
                self.eat();
                self.expect(Token::LParen)?;
                let cond = self.parse_expr()?;
                self.expect(Token::RParen)?;
                let body = self.parse_block()?;
                Ok(Statement::While { cond, body })
            }
            Token::Return => {
                self.eat();
                let expr = self.parse_expr()?;
                Ok(Statement::Return { expr })
            }
            Token::Print => {
                self.eat();
                let expr = self.parse_expr()?;
                Ok(Statement::Print { expr })
            }
            Token::Ident(name)
                if {
                    // could be assignment or function‐call expr
                    let lookahead = &self.tokens.get(self.pos + 1).unwrap_or(&Token::EOF);
                    matches!(lookahead, Token::Eq)
                } =>
            {
                // assignment: ident = expr
                let name = name.clone();
                self.eat();
                self.expect(Token::Eq)?;
                let expr = self.parse_expr()?;
                Ok(Statement::Assign { name, expr })
            }
            _ => {
                // fallback: bare expression statement
                let expr = self.parse_expr()?;
                Ok(Statement::ExprStmt(expr))
            }
        }
    }

    /// Parse expressions with correct precedence:
    /// equality -> comparison -> addition -> term -> factor -> primary
    fn parse_expr(&mut self) -> Result<Expr, CompileError> {
        self.parse_equality()
    }

    fn parse_equality(&mut self) -> Result<Expr, CompileError> {
        let mut lhs = self.parse_comparison()?;
        while matches!(self.peek(), Token::EqEq | Token::Ne) {
            let op = match self.peek() {
                Token::EqEq => BinOp::Eq,
                Token::Ne => BinOp::Ne,
                _ => unreachable!(),
            };
            self.eat();
            let rhs = self.parse_comparison()?;
            lhs = Expr::Binary {
                op,
                left: Box::new(lhs),
                right: Box::new(rhs),
            };
        }
        Ok(lhs)
    }

    fn parse_comparison(&mut self) -> Result<Expr, CompileError> {
        let mut lhs = self.parse_addition()?;
        while matches!(self.peek(), Token::Lt | Token::Le | Token::Gt | Token::Ge) {
            let op = match self.peek() {
                Token::Lt => BinOp::Lt,
                Token::Le => BinOp::Le,
                Token::Gt => BinOp::Gt,
                Token::Ge => BinOp::Ge,
                _ => unreachable!(),
            };
            self.eat();
            let rhs = self.parse_addition()?;
            lhs = Expr::Binary {
                op,
                left: Box::new(lhs),
                right: Box::new(rhs),
            };
        }
        Ok(lhs)
    }

    fn parse_addition(&mut self) -> Result<Expr, CompileError> {
        let mut lhs = self.parse_term()?;
        while matches!(self.peek(), Token::Plus | Token::Minus) {
            let op = if *self.peek() == Token::Plus {
                BinOp::Add
            } else {
                BinOp::Sub
            };
            self.eat();
            let rhs = self.parse_term()?;
            lhs = Expr::Binary {
                op,
                left: Box::new(lhs),
                right: Box::new(rhs),
            };
        }
        Ok(lhs)
    }

    fn parse_term(&mut self) -> Result<Expr, CompileError> {
        let mut lhs = self.parse_factor()?;
        while matches!(self.peek(), Token::Star | Token::Slash | Token::Percent) {
            let op = match self.peek() {
                Token::Star => BinOp::Mul,
                Token::Slash => BinOp::Div,
                Token::Percent => BinOp::Rem,
                _ => unreachable!(),
            };
            self.eat();
            let rhs = self.parse_factor()?;
            lhs = Expr::Binary {
                op,
                left: Box::new(lhs),
                right: Box::new(rhs),
            };
        }
        Ok(lhs)
    }

    fn parse_factor(&mut self) -> Result<Expr, CompileError> {
        match self.peek() {
            Token::Plus => {
                self.eat();
                let e = self.parse_factor()?;
                Ok(Expr::Unary {
                    op: UnOp::Pos,
                    expr: Box::new(e),
                })
            }
            Token::Minus => {
                self.eat();
                let e = self.parse_factor()?;
                Ok(Expr::Unary {
                    op: UnOp::Neg,
                    expr: Box::new(e),
                })
            }
            Token::Number(n) => {
                let v = *n;
                self.eat();
                Ok(Expr::Number(v))
            }
            Token::BoolLiteral(b) => {
                let v = *b;
                self.eat();
                Ok(Expr::Bool(v))
            }
            Token::StrLiteral(s) => {
                let v = s.clone();
                self.eat();
                Ok(Expr::StrLiteral(v))
            }
            Token::Ident(name) => {
                let name = name.clone();
                self.eat();
                // function call?
                if *self.peek() == Token::LParen {
                    self.eat();
                    let mut args = Vec::new();
                    if *self.peek() != Token::RParen {
                        loop {
                            args.push(self.parse_expr()?);
                            if *self.peek() == Token::Comma {
                                self.eat();
                                continue;
                            }
                            break;
                        }
                    }
                    self.expect(Token::RParen)?;
                    Ok(Expr::Call { name, args })
                } else {
                    Ok(Expr::Variable(name))
                }
            }
            Token::LParen => {
                self.eat();
                let e = self.parse_expr()?;
                self.expect(Token::RParen)?;
                Ok(e)
            }
            other => Err(CompileError::Parse(format!(
                "Unexpected token in factor: {:?}",
                other
            ))),
        }
    }
}

File: ocaml/lib/parser.ml
open Ast
open Error
open Lexer

type parser = {
  tokens: token list;
  mutable pos: int;
}

let make_parser tokens = { tokens; pos = 0 }

let peek p = List.nth p.tokens p.pos
let eat p = if p.pos < List.length p.tokens then p.pos <- p.pos + 1
let expect p expected =
  if peek p = expected then (eat p; Ok ())
  else Error (Parse (Printf.sprintf "Expected %s, found %s" (string_of_token expected) (string_of_token (peek p))))


(* helper *)
let rec parse_many_until p stop_token parse_one acc =
  if peek p = stop_token then (
    eat p;
    Ok (List.rev acc)
  ) else 
    match parse_one p with
    | Error e -> Error e
    | Ok item ->
      if peek p = Semicolon then eat p;
      parse_many_until p stop_token parse_one (item :: acc)


let rec parse_program p =
  let rec collect_fns acc =
    if peek p = Fn then 
      match parse_function p with 
      | Error e -> Error e 
      | Ok fn -> collect_fns (fn :: acc)
    else 
      Ok (List.rev acc)
  in
  match collect_fns [] with 
  | Error e -> Error e 
  | Ok functions -> 
    match parse_many_until p EOF parse_statement [] with
    | Error e -> Error e
    | Ok statements -> Ok {functions; statements}


and parse_function p =
  match expect p Fn with
  | Ok () ->
      let name = match peek p with Ident n -> n | _ -> raise (CompileError (Parse "Expected function name")) in
      eat p;
      (match expect p LParen with
       | Ok () ->
           let params = parse_params p [] in
           (match expect p RParen with
            | Ok () ->
                (match parse_block p with
                 | Ok body -> Ok { name; params; body }
                 | Error e -> Error e)
            | Error e -> Error e)
       | Error e -> Error e)
  | Error e -> Error e

and parse_params p acc =
  if peek p = RParen then List.rev acc
  else
    let name = match peek p with Ident n -> n | _ -> raise (CompileError (Parse "Expected parameter name")) in
    eat p;
    let acc' = name :: acc in
    if peek p = Comma then (eat p; parse_params p acc')
    else acc'

and parse_block p =
  match expect p LBrace with
  | Ok () ->
       (match parse_many_until p RBrace parse_statement [] with
       | Error e -> Error e
       | Ok stmts -> Ok stmts
       )
  | Error e -> Error e

and parse_statement p =
  match peek p with
  | Var ->
      eat p;
      let name = match peek p with Ident n -> n | _ -> raise (CompileError (Parse "Expected var name")) in
      eat p;
      (match expect p Eq with
       | Ok () ->
           (match parse_expr p with
            | Ok expr -> Ok (VarDecl (name, expr))
            | Error e -> Error e)
       | Error e -> Error e)
  | Let ->
      eat p;
      let name = match peek p with Ident n -> n | _ -> raise (CompileError (Parse "Expected let name")) in
      eat p;
      (match expect p Eq with
       | Ok () ->
           (match parse_expr p with
            | Ok expr -> Ok (LetDecl (name, expr))
            | Error e -> Error e)
       | Error e -> Error e)
  | If ->
      eat p;
      (match expect p LParen with
       | Ok () ->
           (match parse_expr p with
            | Ok cond ->
                (match expect p RParen with
                 | Ok () ->
                     (match parse_block p with
                      | Ok then_branch ->
                          let else_branch =
                            if peek p = Else then (
                              eat p;
                              match parse_block p with
                              | Ok block -> Some block
                              | Error _ -> None
                            ) else None
                          in
                          Ok (If (cond, then_branch, else_branch))
                      | Error e -> Error e)
                 | Error e -> Error e)
            | Error e -> Error e)
       | Error e -> Error e)
  | While ->
      eat p;
      (match expect p LParen with
       | Ok () ->
           (match parse_expr p with
            | Ok cond ->
                (match expect p RParen with
                 | Ok () ->
                     (match parse_block p with
                      | Ok body -> Ok (While (cond, body))
                      | Error e -> Error e)
                 | Error e -> Error e)
            | Error e -> Error e)
       | Error e -> Error e)
  | Return ->
      eat p;
      (match parse_expr p with
       | Ok expr -> Ok (Return expr)
       | Error e -> Error e)
  | Print ->
      eat p;
      (match parse_expr p with
       | Ok expr -> Ok (Print expr)
       | Error e -> Error e)
  | Ident _ ->
      (match parse_expr p with
       | Ok expr ->
           if peek p = Eq then (
             eat p;
             match parse_expr p with
             | Ok value ->
                 (match expr with
                  | Variable name -> Ok (Assign (name, value))
                  | Index (array, index) -> Ok (IndexAssign (array, index, value))
                  | _ -> Error (Parse "Left-hand side of assignment must be a variable or array index"))
             | Error e -> Error e
           ) else Ok (ExprStmt expr)
       | Error e -> Error e)
  | _ ->
      (match parse_expr p with
       | Ok expr -> Ok (ExprStmt expr)
       | Error e -> Error e)

and parse_expr p = parse_equality p

and parse_equality p =
  match parse_comparison p with
  | Ok lhs ->
      let rec loop lhs =
        match peek p with
        | EqEq ->
            eat p;
            (match parse_comparison p with
             | Ok rhs -> loop (Binary (Eq, lhs, rhs))
             | Error e -> Error e)
        | Ne ->
            eat p;
            (match parse_comparison p with
             | Ok rhs -> loop (Binary (Ne, lhs, rhs))
             | Error e -> Error e)
        | _ -> Ok lhs
      in
      loop lhs
  | Error e -> Error e

and parse_comparison p =
  match parse_addition p with
  | Ok lhs ->
      let rec loop lhs =
        match peek p with
        | Lt ->
            eat p;
            (match parse_addition p with
             | Ok rhs -> loop (Binary (Lt, lhs, rhs))
             | Error e -> Error e)
        | Le ->
            eat p;
            (match parse_addition p with
             | Ok rhs -> loop (Binary (Le, lhs, rhs))
             | Error e -> Error e)
        | Gt ->
            eat p;
            (match parse_addition p with
             | Ok rhs -> loop (Binary (Gt, lhs, rhs))
             | Error e -> Error e)
        | Ge ->
            eat p;
            (match parse_addition p with
             | Ok rhs -> loop (Binary (Ge, lhs, rhs))
             | Error e -> Error e)
        | _ -> Ok lhs
      in
      loop lhs
  | Error e -> Error e

and parse_addition p =
  match parse_term p with
  | Ok lhs ->
      let rec loop lhs =
        match peek p with
        | Plus ->
            eat p;
            (match parse_term p with
             | Ok rhs -> loop (Binary (Add, lhs, rhs))
             | Error e -> Error e)
        | Minus ->
            eat p;
            (match parse_term p with
             | Ok rhs -> loop (Binary (Sub, lhs, rhs))
             | Error e -> Error e)
        | _ -> Ok lhs
      in
      loop lhs
  | Error e -> Error e

and parse_term p =
  match parse_factor p with
  | Ok lhs ->
      let rec loop lhs =
        match peek p with
        | Star ->
            eat p;
            (match parse_factor p with
             | Ok rhs -> loop (Binary (Mul, lhs, rhs))
             | Error e -> Error e)
        | Slash ->
            eat p;
            (match parse_factor p with
             | Ok rhs -> loop (Binary (Div, lhs, rhs))
             | Error e -> Error e)
        | Percent ->
            eat p;
            (match parse_factor p with
             | Ok rhs -> loop (Binary (Rem, lhs, rhs))
             | Error e -> Error e)
        | _ -> Ok lhs
      in
      loop lhs
  | Error e -> Error e

and parse_factor p =
  match peek p with
  | LBracket ->
      eat p;
      let rec parse_elems acc =
        if peek p = RBracket then Ok (List.rev acc)
        else
          match parse_expr p with
          | Ok e ->
              let acc' = e :: acc in
              if peek p = Comma then (eat p; parse_elems acc')
              else parse_elems acc'
          | Error e -> Error e
      in
      (match parse_elems [] with
       | Ok elems ->
           (match expect p RBracket with
            | Ok () -> Ok (ArrayLiteral elems)
            | Error e -> Error e)
       | Error e -> Error e)
  | Number n ->
      eat p;
      Ok (Number n)
  | BoolLiteral b ->
      eat p;
      Ok (Bool b)
  | StrLiteral s ->
      eat p;
      Ok (StrLiteral s)
  | Ident name ->
      eat p;
      if peek p = LParen then (
        eat p;
        let rec parse_args acc =
          if peek p = RParen then Ok (List.rev acc)
          else
            match parse_expr p with
            | Ok arg ->
                let acc' = arg :: acc in
                if peek p = Comma then (eat p; parse_args acc')
                else parse_args acc'
            | Error e -> Error e
        in
        (match parse_args [] with
         | Ok args ->
             (match expect p RParen with
              | Ok () -> Ok (Call (name, args))
              | Error e -> Error e)
         | Error e -> Error e)
      ) else Ok (Variable name)
  | LParen ->
      eat p;
      (match parse_expr p with
       | Ok e ->
           (match expect p RParen with
            | Ok () -> Ok e
            | Error e -> Error e)
       | Error e -> Error e)
  | other ->
      Error (Parse (Printf.sprintf "Unexpected token in factor: %s" (string_of_token other)))